File: 20221226_Monday/20221226.go
// Leetcode: 55. Jump Game.
// https://leetcode.com/problems/jump-game/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 56 ms, faster than 97.44% of Go online submissions for Jump Game.
// Memory Usage: 6.9 MB, less than 91.99% of Go online submissions for Jump Game.
// 2022.12.26 Daily Challenge.
func canJump(nums []int) bool {
  l := len(nums)-1
  j := 0 // Farest index player could reach.
  for i,v := range nums {
    if j < i { return false}
    if j < i + v { j = i + v}
    if j >= l { return true}
  }
  return j >= l
}

File: 20221226_Monday/20221226.rb
# Leetcode: 55. Jump Game.
# https://leetcode.com/problems/jump-game/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 163 ms, faster than 70.77% of Ruby online submissions for Jump Game.
# Memory Usage: 216 MB, less than 87.69% of Ruby online submissions for Jump Game.
# 2022.12.26 Daily Challenge.
# @param {Integer[]} nums
# @return {Boolean}
def can_jump(nums)
  l = nums.size - 1
  j = 0 # j is maximum index player could reach.
  nums.each_with_index do |val,i|
    return false if j < i # Could not reach.
    j = i + val if j < i + val # Push j further.
    return true if j >= l # It is possible to reach last index.
  end
  raise "This could not be reached"
end
